BUG: Fix display node copy in CloneVolume

agirault opened the pull request:
**Motivation**
If a volume is rendered through the Volume Rendering volume module then
cropped through the Crop Volume module, it is not possible to display the
volume rendering of the cropped volume: the original volume would be
displayed instead.

**Reason**
In the Crop Volume module, the method `vtkSlicerVolumesLogic::CloneVolume`
was called.  The volume and its first display node were cloned using
`vtkMRMLVolumeNode::CopyWithScene`, but the remaining display nodes were
ignored. This was resulting in having a cloned volume with references to
the original volume display nodes. In the case above, the display node of
type `vtkMRMLVolumeRenderingDisplayNode` was the same for the cropped and
original volumes, referencing the original volume for both of them.

**Fix**
In this commit only the first display node is cloned, and the other display
nodes are removed. Cloning a `vtkMRMLVolumeRenderingDisplayNode` would not be
as straightforward as copying a `vtkMRMLVolumeDisplayNode` since the display
node itself has a `VolumeNodeID`, a `ROINodeID` and a `VolumePropertyNodeID` that
would need to be created/updated.

This fixes the issue 4109: http://na-mic.org/Mantis/view.php?id=4109

jcfr commented on Modules/Loadable/Volumes/Logic/vtkSlicerVolumesLogic.cxx:
```diff
@@ -1216,15 +1204,38 @@ CloneVolume (vtkMRMLScene *scene, vtkMRMLVolumeNode *volumeNode, const char *nam
-  if ( clonedDisplayNode )
+
+  // remove display nodes (but not the first one)
+  for ( unsigned int i = 1; i < clonedVolumeNode->GetNumberOfDisplayNodes(); ++i )
```
Nitpicks: This form may be simpler:

```
while (clonedVolumeNode->GetNumberOfDisplayNodes() > 1)
  {
  clonedVolumeNode->RemoveNthDisplayNodeID(1); // always remove at index 1 since they will shift
  }
```

jcfr commented:
Thanks for the patch. Looks good. :+1: 

Could you also add/update test of the Volume logic ?

fedorov commented:
Thank you for working on this!

pieper commented:
Agreed - thanks for the fix!

On Wed, Jan 6, 2016 at 5:44 PM, Andrey Fedorov <notifications@github.com>
wrote:

> Thank you for working on this!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Slicer/Slicer/pull/439#issuecomment-169486224>.
>

agirault commented:
I'll work on the test for the Volume logic now.

agirault commented:
I refactored the `vtkSlicerVolumesLogicTest1` file and added new tests for `vtkSlicerVolumesLogic::CloneVolume` which check :
1. That a cloned volume has no storage node.
2. That a cloned volume has display nodes not referencing the original volume (fixed in commit 6ed6752 by copying only the first display node and removing the others).
3. That a cloned volume has the same imageData than the original volume.

Testing the copy of the display nodes involved including `vtkSlicerVolumeRenderingModuleMRML` in the CMakeLists.txt to be able to test the `vtkMRMLVolumeRenderingDisplayNode` classes. I don't think the way I did it is correct so please let me know how to make this cleaner.

Thank you.

jcfr commented:
> I don't think the way I did it is correct so please let me know

Looks good I will integrate asap.

jcfr commented:
@agirault Would it be possible to base the last two commits on top of the current master. There are merge conflicts.

Note that tests and associated functions should now all return EXIT_FAILURE, EXIT_SUCCESS 

Thanks

agirault commented:
Rebase made.

As for the return values, isn't it confusing since `true` will be equivalent to `EXIT_FAILURE` (1) and `false` equivalent to `EXIT_SUCCESS` (0) ?

jcfr commented:
Thanks.

That is the new convention. Since function returns int and use the EXIT_.
It is okay.
On Jan 8, 2016 3:26 PM, "Alexis Girault" <notifications@github.com> wrote:

> Rebase made.
>
> As for the return values, isn't it confusing since true will be
> equivalent to EXIT_FAILURE (1) and false equivalent to EXIT_SUCCESS (0) ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Slicer/Slicer/pull/439#issuecomment-170114149>.
>

jcfr commented:
First two commits integrated as [r24861](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=24861) and [r24862](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=24862)

agirault commented:
@jcfr you integrated the second commit which tests returning `bool` instead of `int`. I am supposed to integrate the testing macros in the latest commit with the testing of my changes?

jcfr commented:
> integrated the second commit which tests returning bool instead of int

Indeed. This is my mistake.

> I am supposed to integrate the testing macros in the latest commit with the testing of my changes

Would be great if you could update things to use `EXIT_*` and `int` in your next commit.

Thanks

agirault commented:
Sorry for the delay, I had been busy on another project. I'll do this now. Since you integrated the two first commits what would you say would be the best way to update the last commit @jcfr ? Should I just cherry-pick this last commit from master, correct the conflicts and update the return values, then push force to only do the PR for this commit?

agirault commented:
Sorry for the delay, this should be good now @jcfr.

jcfr commented:
Thanks for following up with your contribution :+1: 

Integrated in [r24890](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=24890)
